### PR TITLE
Corecct ProcessInfo.operationSystemVersion on Windows

### PR DESCRIPTION
**Pull request by yamoridon**
At Windows 8.1, Microsoft changed the default behavior of
GetVersionEx not to return actual version number for preventing
compatibility isseus. I think it's better to use the product version
retrived from kernel32.dll as an alternative as Microsoft described in
a document:
https://docs.microsoft.com/en-us/windows/desktop/SysInfo/getting-the-system-version

### CODE REVIEW

**Comment by compnerd** [comment]
The usage here is not for compatibility but rather reporting purposes.  I agree with @jakepetroules - we should just resort to `RtlGetVersion`

**Comment by yamoridon** [comment]
If my understanding is correct, we can't use that approach. `RtlGetVersion` is a kernel mode function, so application programs can't use it.

**Comment by compnerd** [comment]
@yamoridon - no your understanding is not correct.  The RtlFunctions are implemented in the kernel, but you can still use them in usermode.  It just is a slightly roundabout way - you need to use `GetProcAddress`.  If you want, I can make the change - its pretty easy.

File: Foundation/ProcessInfo.swift [hunk 1]
```diff
@@ -104,12 +104,23 @@ open class ProcessInfo: NSObject {
         }
         versionString = productVersion._swiftObject
 #elseif os(Windows)
-        var siVersionInfo: OSVERSIONINFOW = OSVERSIONINFOW()
-        siVersionInfo.dwOSVersionInfoSize = DWORD(MemoryLayout<OSVERSIONINFOEXW>.size)
-        if GetVersionExW(&siVersionInfo) == FALSE {
-          return OperatingSystemVersion(majorVersion: fallbackMajor, minorVersion: fallbackMinor, patchVersion: fallbackPatch)
+        guard let ntoskrnl = LoadLibraryExA("ntoskrnl.exe", nil, DWORD(LOAD_LIBRARY_SEARCH_SYSTEM32)) else {
```
Comment (compnerd) [inline Foundation/ProcessInfo.swift]: Please use `LoadLibraryW`.

File: Foundation/ProcessInfo.swift [hunk 2]
```diff
@@ -104,12 +104,23 @@ open class ProcessInfo: NSObject {
...
         }
-        return OperatingSystemVersion(majorVersion: Int(siVersionInfo.dwMajorVersion), minorVersion: Int(siVersionInfo.dwMinorVersion), patchVersion: Int(siVersionInfo.dwBuildNumber))
+        var osVersionInfo = RTL_OSVERSIONINFOW()
+        osVersionInfo.dwOSVersionInfoSize = DWORD(MemoryLayout<RTL_OSVERSIONINFOW>.size)
+        guard rtlGetVersion(&osVersionInfo) == 0 else {
+            return OperatingSystemVersion(majorVersion: fallbackMajor, minorVersion: fallbackMinor, patchVersion: fallbackPatch)
+        }
+        let majorVersion = Int(osVersionInfo.dwMajorVersion)
+        let minorVersion = Int(osVersionInfo.dwMinorVersion)
+        let patchVersion = Int(osVersionInfo.dwBuildNumber)
```
Comment (compnerd) [inline Foundation/ProcessInfo.swift]: Just inline the `Int` conversion.

File: Foundation/ProcessInfo.swift [hunk 3]
```diff
@@ -104,12 +104,25 @@ open class ProcessInfo: NSObject {
         }
         versionString = productVersion._swiftObject
 #elseif os(Windows)
-        var siVersionInfo: OSVERSIONINFOW = OSVERSIONINFOW()
-        siVersionInfo.dwOSVersionInfoSize = DWORD(MemoryLayout<OSVERSIONINFOEXW>.size)
-        if GetVersionExW(&siVersionInfo) == FALSE {
-          return OperatingSystemVersion(majorVersion: fallbackMajor, minorVersion: fallbackMinor, patchVersion: fallbackPatch)
+        let libraryName: [WCHAR] = [ 0x6e, 0x74, 0x6f, 0x73, 0x6b, 0x72, 0x6e, 0x6c, 0x2e, 0x65, 0x78, 0x65, 0x00 ] // ntoskrnl.exe
```
Comment (compnerd) [inline Foundation/ProcessInfo.swift]: This should be from NTDLL.dll, not ntoskrnl.exe.  Writing this as:
```
"ntdll.dll".withCString(as: .utf16) {
  guard let ntdll = LoadLibraryExW($0, nil, DWORD(LOAD_LIBRARY_SEARCH_SYSTEM32)) else {
    ...
  }
  ...
}
```

is much more readable.

**Comment by yamoridon** [comment]
Thank you for advice. How about this? Too long line?

File: Foundation/ProcessInfo.swift [hunk 4]
```diff
@@ -104,12 +104,24 @@ open class ProcessInfo: NSObject {
         }
         versionString = productVersion._swiftObject
 #elseif os(Windows)
-        var siVersionInfo: OSVERSIONINFOW = OSVERSIONINFOW()
-        siVersionInfo.dwOSVersionInfoSize = DWORD(MemoryLayout<OSVERSIONINFOEXW>.size)
-        if GetVersionExW(&siVersionInfo) == FALSE {
-          return OperatingSystemVersion(majorVersion: fallbackMajor, minorVersion: fallbackMinor, patchVersion: fallbackPatch)
+        guard let ntdll = "ntdll.dll".withCString(encodedAs: UTF16.self, { LoadLibraryExW($0, nil, DWORD(LOAD_LIBRARY_SEARCH_SYSTEM32)) }) else {
```
Comment (compnerd) [inline Foundation/ProcessInfo.swift]: Make the closure trailing, the `guard let` and add a newline after the brace perhaps?

```suggestion
        guard let ntdll = "ntdll.dll".withCString(encodedAs: UTF16.self) {
          LoadLibraryExW($0, nil, DWORD(LOAD_LIBRARY_SEARCH_SYSTEM32))
        } else {
```

File: Foundation/ProcessInfo.swift [hunk 5]
```diff
@@ -104,12 +104,24 @@ open class ProcessInfo: NSObject {
...
+            return OperatingSystemVersion(majorVersion: fallbackMajor, minorVersion: fallbackMinor, patchVersion: fallbackPatch)
+        }
+        defer { FreeLibrary(ntdll) }
+        typealias RtlGetVersion = @convention(c) (UnsafeMutablePointer<RTL_OSVERSIONINFOW>) -> NTSTATUS
+        guard let rtlGetVersion = unsafeBitCast(GetProcAddress(ntdll, "RtlGetVersion"), to: Optional<RtlGetVersion>.self) else {
```
Comment (compnerd) [inline Foundation/ProcessInfo.swift]: Can you name this `pfnRTLGetVersion` instead please?  It makes it easier to find the references when the names match the actual function name.

File: Foundation/ProcessInfo.swift [hunk 6]
```diff
@@ -104,12 +104,24 @@ open class ProcessInfo: NSObject {
...
 #elseif os(Windows)
-        var siVersionInfo: OSVERSIONINFOW = OSVERSIONINFOW()
-        siVersionInfo.dwOSVersionInfoSize = DWORD(MemoryLayout<OSVERSIONINFOEXW>.size)
-        if GetVersionExW(&siVersionInfo) == FALSE {
-          return OperatingSystemVersion(majorVersion: fallbackMajor, minorVersion: fallbackMinor, patchVersion: fallbackPatch)
+        guard let ntdll = "ntdll.dll".withCString(encodedAs: UTF16.self, { LoadLibraryExW($0, nil, DWORD(LOAD_LIBRARY_SEARCH_SYSTEM32)) }) else {
+            return OperatingSystemVersion(majorVersion: fallbackMajor, minorVersion: fallbackMinor, patchVersion: fallbackPatch)
+        }
+        defer { FreeLibrary(ntdll) }
+        typealias RtlGetVersion = @convention(c) (UnsafeMutablePointer<RTL_OSVERSIONINFOW>) -> NTSTATUS
```
Comment (compnerd) [inline Foundation/ProcessInfo.swift]: Make this `RTLGetVersionTy`

**Comment by compnerd** [comment]
Minor naming things, but looks good otherwise!

**Comment by compnerd** [comment]
@swift-ci please test

File: Foundation/ProcessInfo.swift (hunk 4 above)
Comment (yamoridon) [reply Foundation/ProcessInfo.swift]: Thank you for your advice again. Your suggestion's form requires additonal parentheses before `"ntdll.dll"` and `else` like this:
```swift
        guard let ntdll = ("ntdll.dll".withCString(encodedAs: UTF16.self) {
          LoadLibraryExW($0, nil, DWORD(LOAD_LIBRARY_SEARCH_SYSTEM32))
        }) else {
            return OperatingSystemVersion(majorVersion: fallbackMajor, minorVersion: fallbackMinor, patchVersion: fallbackPatch)
        }
```
Without those, the compiler gets confused.
```
C:/swift/swift-corelibs-foundation/Foundation/ProcessInfo.swift:107:74: error: expected 'else' after 'guard' condition
        guard let ntdll = "ntdll.dll".withCString(encodedAs: UTF16.self) {
                                                                         ^
                                                                         else
C:/swift/swift-corelibs-foundation/Foundation/ProcessInfo.swift:108:26: error: anonymous closure argument not contained in a closure
          LoadLibraryExW($0, nil, DWORD(LOAD_LIBRARY_SEARCH_SYSTEM32))
                         ^
C:/swift/swift-corelibs-foundation/Foundation/ProcessInfo.swift:109:10: error: consecutive statements on a line must be separated by ';'
        } else {
         ^
         ;
C:/swift/swift-corelibs-foundation/Foundation/ProcessInfo.swift:109:11: error: expected expression
        } else {
          ^
C:/swift/swift-corelibs-foundation/Foundation/ProcessInfo.swift:107:39: error: cannot invoke 'withCString' with an argument list of type '(encodedAs: UTF16.Type)'
        guard let ntdll = "ntdll.dll".withCString(encodedAs: UTF16.self) {
                                      ^
C:/swift/swift-corelibs-foundation/Foundation/ProcessInfo.swift:109:16: error: closure expression is unused
        } else {
               ^
C:/swift/swift-corelibs-foundation/Foundation/ProcessInfo.swift:109:16: note: did you mean to use a 'do' statement?
        } else {
               ^
               do
```
Do you prefer your suggestion with additonal parentheses or a form in PR?

Comment (compnerd) [reply Foundation/ProcessInfo.swift]: The wrapped form is definitely easier to read IMO.

**Comment by compnerd** [comment]
Can you also please squash the changes?

**Comment by yamoridon** [comment]
Does it mean following?

1. Close this PR.
2. Make a new branch.
3. Make changes on this PR as one commit on the new branch.
4. Make a new PR

**Comment by compnerd** [comment]
You can use `git rebase -i` to squash all the commits into a single commit and force push to the existing branch :-).

**Comment by yamoridon** [comment]
Seems to  have succeeded. Thank you for your kind support!

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
I think that this improves the Windows status quo, so I'm going to go ahead and merge this.  If @millenomi has additional suggestions, they can be handled in a follow up.

**Pull request merged**